docs: updates for CREATE FUNCTION attributes EXECUTE ON

mkiyama opened the pull request:
updated docs for EXECUTE ON attributes.
Changes with links to HTML on GPDB review site.

Updated EXECUTE ON attributes in:
[CREATE FUNCTION](http://docs-gpdb-review-staging.cfapps.io/540/ref_guide/sql_commands/CREATE_FUNCTION.html)
[ALTER FUNCTION](http://docs-gpdb-review-staging.cfapps.io/540/ref_guide/sql_commands/ALTER_FUNCTION.html)

Added statement that volatility and EXECUTE ON attributes not supported in DO command.
[DO command](http://docs-gpdb-review-staging.cfapps.io/540/ref_guide/sql_commands/DO.html)

Incorporated EXECUTE ON information to Functions and Operators. 
[Functions and Operators](http://docs-gpdb-review-staging.cfapps.io/540/admin_guide/query/topics/functions-operators.html)

Added EXECUTE ON limitation to GPORCA Limitations.

[GPORCA Limitations](http://docs-gpdb-review-staging.cfapps.io/540/admin_guide/query/topics/query-piv-opt-limitations.html)

Updated PL/Python, PL/Perl examples to use attributes.
[PL/Perl](http://docs-gpdb-review-staging.cfapps.io/540/ref_guide/extensions/pl_perl.html)
[PL/Python](http://docs-gpdb-review-staging.cfapps.io/540/ref_guide/extensions/pl_python.html)

Updated g_proc table - added columns: proconfig, prodataaccess, proexeclocation
[pg_proc table](http://docs-gpdb-review-staging.cfapps.io/540/ref_guide/system_catalogs/pg_proc.html)

see commit
https://github.com/greenplum-db/gpdb/commit/aa148d2a3cba9866a2c3a06364a26dee32d1c0a2

ivannovick commented:
hi guys, i lost track is this capability on GP 5 or GP 6?

ivannovick commented:
any chance we can have examples in the docs for EXECUTE ON?

mkiyama commented:
This is GP 6. There is a catalog change.
There is a simple example in CREATE FUNCTION

ivannovick commented:
Thx Mel, this is great, we should call this out in the new features for GP 6 release notes.  I think its important.

mkiyama commented:
@hlinnaka 
could you take a look at this PR?
Major changes are in CREATE FUNCTION and pg_proc table.

Also added that these attributes are not supported in DO command. 

Updated EXECUTE ON attributes in:
[CREATE FUNCTION](http://docs-gpdb-review-staging.cfapps.io/540/ref_guide/sql_commands/CREATE_FUNCTION.html)

Updated pg_proc table - added columns: proconfig, prodataaccess, proexeclocation
[pg_proc table](http://docs-gpdb-review-staging.cfapps.io/540/ref_guide/system_catalogs/pg_proc.html)

Thanks.

mkiyama commented:
@hlinnaka 
Have you had a chance to review this?

hlinnaka commented on gpdb-doc/dita/ref_guide/sql_commands/DO.xml:
```diff
@@ -58,12 +58,15 @@
       <p>The user must have <codeph>USAGE</codeph> privilege for the procedural language, or must be
         a superuser if the language is untrusted. This is the same privilege requirement as for
         creating a function in the language.</p>
+      <p>Anonymous blocks do not support function volatility or <codeph>EXECUTE ON</codeph>
```
Might be good to explicitly mention that a DO block is always executed on the master.

hlinnaka commented:
On 10/01/18 00:22, Mel Kiyama wrote:
> @hlinnaka
> Have you had a chance to review this?

Sorry, [excuses here]. Reviewed now, looks good to me! There were a lot 
of formatting changes of the sources, which made it a bit tedious to see 
what the actual changes were - would be nice to do those as a separate 
commit.

- Heikki

mkiyama commented:
sorry about that. I'll keep that in mind.

dyozie commented on gpdb-doc/dita/admin_guide/query/topics/functions-operators.xml:
```diff
@@ -133,6 +194,13 @@ SELECT foo();
                 default, user-defined functions are declared as <codeph>VOLATILE</codeph>, so if
                 your user-defined function is <codeph>IMMUTABLE</codeph> or <codeph>STABLE</codeph>,
                 you must specify the correct volatility level when you register your function.</p>
+            <p>By default, user-defined functions are declared as <codeph>EXEUCTE ON ANY</codeph>. A
```
Typo:  EXEUCTE

dyozie commented on gpdb-doc/dita/admin_guide/query/topics/functions-operators.xml:
```diff
@@ -133,6 +194,13 @@ SELECT foo();
                 you must specify the correct volatility level when you register your function.</p>
+            <p>By default, user-defined functions are declared as <codeph>EXEUCTE ON ANY</codeph>. A
+                function that executes queries to access tables is supported only when a function is
+                executed on the master instance. Such functions should be defined with the
```
Probably change "should" to "must" here.

dyozie commented on gpdb-doc/dita/admin_guide/query/topics/functions-operators.xml:
```diff
@@ -26,16 +26,32 @@
+            <p>For example, a function defined with the <codeph>IMMUTABLE</codeph> attribute can be
+                executed at query planning time, while a function with the <codeph>VOLATILE</codeph>
+                attribute must be executed for every row in the query. A function with the
+                    <codeph>EXECUTE ON MASTER</codeph> attribute is executed only on the master
```
I think some passive voice makes sense when describing this new functionality (like in the previous para).  But would like to remove it in other instances.  Here, I think it would be clearer to say:  "A function with the EXECUTE ON MASTER attribute executes only on the master segment, ..."

dyozie commented on gpdb-doc/dita/admin_guide/query/topics/functions-operators.xml:
```diff
@@ -26,16 +26,32 @@
+                attribute must be executed for every row in the query. A function with the
+                    <codeph>EXECUTE ON MASTER</codeph> attribute is executed only on the master
+                segment, and a function with the <codeph>EXECUTE ON ALL SEGMENTS</codeph> attribute
+                is executed on all primary segment instances (not the master). </p>
```
"is executed" -> "executes"

dyozie commented on gpdb-doc/dita/admin_guide/query/topics/functions-operators.xml:
```diff
@@ -66,19 +82,66 @@
+                            <entry colname="col1">EXECUTE ON ANY</entry>
+                            <entry colname="col3">Indicates that the function can be executed on the
+                                master, or any segment instance, and it returns the same result
+                                regardless of where it is executed. This is the default
```
"is executed" -> "executes"

dyozie commented on gpdb-doc/dita/admin_guide/query/topics/functions-operators.xml:
```diff
@@ -66,19 +82,66 @@
+                                regardless of where it is executed. This is the default
+                                attribute.</entry>
+                            <entry colname="col4">Greenplum Database determines where the function
+                                is executed.</entry>
```
"is executed" -> "executes"

dyozie commented on gpdb-doc/dita/admin_guide/query/topics/functions-operators.xml:
```diff
@@ -133,6 +194,13 @@ SELECT foo();
                 your user-defined function is <codeph>IMMUTABLE</codeph> or <codeph>STABLE</codeph>,
                 you must specify the correct volatility level when you register your function.</p>
+            <p>By default, user-defined functions are declared as <codeph>EXEUCTE ON ANY</codeph>. A
+                function that executes queries to access tables is supported only when a function is
```
"is executed" -> "executes"

dyozie commented on gpdb-doc/dita/ref_guide/sql_commands/ALTER_FUNCTION.xml:
```diff
@@ -11,80 +16,116 @@ ALTER FUNCTION <varname>name</varname> ( [ [<varname>argmode</varname>] [<varnam
+          <pt>EXECUTE ON MASTER</pt>
+          <pt>EXECUTE ON ALL SEGMENTS</pt>
+          <pd>The <codeph>EXECUTE ON</codeph> attributes specify where (master or segment instance)
+            a function is executed when it is invoked during the query execution process.</pd>
```
"is executed" -> "executes"

dyozie commented on gpdb-doc/dita/ref_guide/sql_commands/ALTER_FUNCTION.xml:
```diff
@@ -11,80 +16,116 @@ ALTER FUNCTION <varname>name</varname> ( [ [<varname>argmode</varname>] [<varnam
+          <pd><codeph>EXECUTE ON ANY</codeph> (the default) indicates that the function can be
+            executed on the master, or any segment instance, and it returns the same result
+            regardless of where it is executed. Greenplum Database determines where the function is
+            executed.</pd>
```
"is executed" -> "executes" (last instance only)

dyozie commented on gpdb-doc/dita/ref_guide/sql_commands/ALTER_FUNCTION.xml:
```diff
@@ -11,80 +16,116 @@ ALTER FUNCTION <varname>name</varname> ( [ [<varname>argmode</varname>] [<varnam
+            executed on the master, or any segment instance, and it returns the same result
+            regardless of where it is executed. Greenplum Database determines where the function is
+            executed.</pd>
+          <pd><codeph>EXECUTE ON MASTER</codeph> indicates that the function must be executed on the
```
"must be executed" -> "must execute only"

dyozie commented on gpdb-doc/dita/ref_guide/sql_commands/ALTER_FUNCTION.xml:
```diff
@@ -11,80 +16,116 @@ ALTER FUNCTION <varname>name</varname> ( [ [<varname>argmode</varname>] [<varnam
+            executed.</pd>
+          <pd><codeph>EXECUTE ON MASTER</codeph> indicates that the function must be executed on the
+            master instance. </pd>
+          <pd><codeph>EXECUTE ON ALL SEGMENTS</codeph> indicates that the function must be executed
```
"be executed" -> "execute"

dyozie commented on gpdb-doc/dita/ref_guide/sql_commands/CREATE_FUNCTION.xml:
```diff
@@ -17,154 +22,225 @@
+                        <codeph>IMMUTABLE</codeph> attribute can be executed at query planning time,
+                    while a function with the <codeph>VOLATILE</codeph> attribute must be executed
+                    for every row in the query. A function with the <codeph>EXECUTE ON
+                        MASTER</codeph> attribute is executed only on the master segment and a
```
Same edits from function-operators above.

dyozie commented on gpdb-doc/dita/ref_guide/sql_commands/CREATE_FUNCTION.xml:
```diff
@@ -17,154 +22,225 @@
+                    <pt>EXECUTE ON MASTER</pt>
+                    <pt>EXECUTE ON ALL SEGMENTS</pt>
+                    <pd>The <codeph>EXECUTE ON</codeph> attributes specify where (master or segment
+                        instance) a function is executed when it is invoked during the query
```
Same edits from above.

dyozie commented on gpdb-doc/dita/ref_guide/sql_commands/CREATE_FUNCTION.xml:
```diff
@@ -17,154 +22,225 @@
+                    Greenplum Database.</p></sectiondiv>
+            <sectiondiv><b>Function Volitionality and EXECUTE ON Attributes</b><p>Volatility
+                    attributes (<codeph>IMMUTABLE</codeph>, <codeph>STABLE</codeph>,
+                        <codeph>VOLITILE</codeph>) and <codeph>EXECUTE ON</codeph> attributes
```
Typoe:  VOLITILE

dyozie commented on gpdb-doc/dita/ref_guide/sql_commands/CREATE_FUNCTION.xml:
```diff
@@ -278,21 +403,42 @@ CREATE FUNCTION return_enum_as_array( anyenum, anyelement, anyelement )
+  begin 
+    return next ($1 || ' - ' || now()::text ); 
+  end;
+ $$ language plpgsql VOLITILE EXECUTE ON ALL SEGMENTS;
```
Typo: VOLITILE

dyozie commented on gpdb-doc/dita/ref_guide/system_catalogs/pg_proc.xml:
```diff
@@ -258,6 +264,23 @@
+            <entry colname="col1"><codeph>proexeclocation</codeph></entry>
+            <entry colname="col2">char</entry>
+            <entry colname="col3"/>
+            <entry colname="col4">Where the function is executed when it is invoked:
```
"is executed" -> executes
